README.md
# Bitcoin P2P Manager
Manage a network of peers in a Bitcoin peer-to-peer network.

## Events
The BTCNetwork object translates the raw network messages received from the PeerManager into parsed objects. Here's the differences between the various events (Events are identified like `Object::EventName`):

* `Peer::message`: `{ peer: Peer, command: STRING, data: Buffer }` Raw message, fired for every message. Bubbled-up as `PeerManager::message`
* `BTCNetwork::message`: `{ peer: Peer, command: STRING, data: { version: INT, services: Buffer, time: INT, ...}` Parsed data
* `Peer::versionMessage`: `{ peer: Peer, data: Buffer }` Raw message. Bubbled-up by PeerManager as `PeerManager::versionMessage`
* `BTCNetwork::versionMessage`: `{ peer: Peer, version: INT, services: Buffer, time: INT, ... }` Parsed data

This is accomplished by several message handlers on the BTCNetwork object, named in the format `parseCommandMessage` (e.g. `parseVersionMessage`, `parseInvMessage`, `parseGetaddrMessage`, etc.). If a handler is not found for a message type, the raw form is bubbled up. To handle additional types of messages, add a handler method named appropriately, and it will be passed `data` and `peer` as arguments. It must return an object; if the result is an array (i.e `inv` messages), return an object with one property set to the array (i.e. `{ items: ARRAY }`).

These parsing methods mean, that if you're subscribed to an event stream that presents raw data (i.e. the `versionMessage` events from an individual Peer), you can parse them like:

```js
var p = new Peer('example.com');
var btc = new BTCNetwork();

p.on('versionMessage', function(d) {
  var parsed = btc.parseVersionMessage(d.data, d.peer);
});
```

CHANGELOG.md
Next Release
------------
* Update the `Struct` object error-handling to not use events, but rather internal failed state flag.

0.2.0 / 2014-01-16
------------------
* Add parsing of more message types: `alert`, `getblocks`, `getheaders`, `headers`, `notfound`, and `tx`
* Separate `parse` from `handle`, so message payloads can be parsed independently of affecting the p2p network management

0.1.1 / 2014-01-13
------------------
* FIX: syntax error in prior release

0.1.0 / 2014-01-10
------------------
* Initial release

test/struct.test.js
var Struct = require('../lib/Struct').Struct;
var assert = require("assert")

describe('Struct', function() {
  it('should fail on increment overflow', function() {
    var s = new Struct(new Buffer(10));
    for (var i = 0; i < 11; i++) {
      s.incrPointer(1);
    }
    assert.equal(s.hasFailed, true);
  });
  it('should fail on set overflow', function() {
    var s = new Struct(new Buffer(10));
    s.setPointer(11);
    assert.equal(s.hasFailed, true);
  });
  it('readInt8() should return sequential 8-bit numbers', function() {
    var s = new Struct(new Buffer([1,2,3,4,5]));
    var test = [];
    for (var i = 0; i < 5; i++) {
      test.push(s.readInt8());
    }
    assert.equal(test.join(','), '1,2,3,4,5');
    assert.equal(s.hasFailed, false);
  });
  it('readUInt16LE() should return sequential 16-bit numbers', function() {
    var s = new Struct(new Buffer([1,0,2,0,3,0,4,0,5,0]));
    var test = [];
    for (var i = 0; i < 5; i++) {
      test.push(s.readUInt16LE());
    }
    assert.equal(test.join(','), '1,2,3,4,5');
    assert.equal(s.hasFailed, false);
  });
  it('readUInt32LE() should return sequential 32-bit numbers', function() {
    var s = new Struct(new Buffer([1,0,0,0,2,0,0,0,3,0,0,0,4,0,0,0,5,0,0,0]));
    var test = [];
    for (var i = 0; i < 5; i++) {
      test.push(s.readUInt32LE());
    }
    assert.equal(test.join(','), '1,2,3,4,5');
    assert.equal(s.hasFailed, false);
  });
  describe('readVarInt()', function() {
    it('should parse 8-bit numbers', function() {
      var s = new Struct(new Buffer([1,2,3,4,5]));
      var test = [];
      for (var i = 0; i < 5; i++) {
        test.push(s.readVarInt());
      }
      assert.equal(test.join(','), '1,2,3,4,5');
      assert.equal(s.hasFailed, false);
    });
    it('should parse 16-bit numbers', function() {
      var s = new Struct(new Buffer([253,1,0,253,2,0,253,3,0,253,4,0,253,5,0]));
      var test = [];
      for (var i = 0; i < 5; i++) {
        test.push(s.readVarInt());
      }
      assert.equal(test.join(','), '1,2,3,4,5');
      assert.equal(s.hasFailed, false);
    });
    it('should parse 32-bit numbers', function() {
      var s = new Struct(new Buffer([254,1,0,0,0,254,2,0,0,0,254,3,0,0,0,254,4,0,0,0,254,5,0,0,0]));
      var test = [];
      for (var i = 0; i < 5; i++) {
        test.push(s.readVarInt());
      }
      assert.equal(test.join(','), '1,2,3,4,5');
      assert.equal(s.hasFailed, false);
    });
    it('should parse 64-bit numbers', function() {
      var s = new Struct(new Buffer([255,1,0,0,0,0,0,0,0,255,2,0,0,0,0,0,0,0,255,3,0,0,0,0,0,0,0,255,4,0,0,0,0,0,0,0,255,5,0,0,0,0,0,0,0]));
      var test = [];
      for (var i = 0; i < 5; i++) {
        var rs = s.readVarInt();
        assert.ok(Buffer.isBuffer(rs));
        assert.equal(rs.length, 8);
        assert.equal(rs.readUInt32LE(4), 0);
        test.push(rs.readUInt32LE(0));
      }
      assert.equal(test.join(','), '1,2,3,4,5');
      assert.equal(s.hasFailed, false);
    });
  });
  it('readVarString() should return proper result', function() {
    var s = new Struct(new Buffer('0F2F5361746F7368693A302E372E322F', 'hex'));
    assert.equal(s.readVarString(), '/Satoshi:0.7.2/');
    assert.equal(s.hasFailed, false);
  });
  it('raw() should return a new Buffer', function() {
    var s = new Struct(new Buffer([1,2,3,4,5]));
    var test = s.raw(3);
    assert.ok(Buffer.isBuffer(test));
    assert.equal(test.length, 3);
    assert.equal(test.toString('hex'), '010203');
    test[1] = 255; // set value in the result
    s.setPointer(1);
    assert.equal(s.readInt8(), 2); // old value should be unaffected
    assert.equal(s.hasFailed, false);
  })
});
 

test/btcnetwork.test.js
var BTCNetwork = require('../lib/BTCNetwork').BTCNetwork;
var assert = require("assert")


var n = new BTCNetwork();
var p = {}; // Dummy Peer

describe('BTCNetwork', function() {
  it('should properly parse an ADDR message', function() {
    var test = n.parseAddrMessage(new Buffer('013722d652010000000000000000000000000000000000ffff525f8c25208d', 'hex'));
    var addrs = test.addrs;
    assert.ok(Array.isArray(addrs));
    assert.equal(addrs.length, 1);
    assert.equal(addrs[0].timestamp.getTime(), 1389765175000);
    assert.equal(addrs[0].services.readUInt32LE(0), 1);
    assert.equal(addrs[0].host, '82.95.140.37');
    assert.equal(addrs[0].port, 8333);
  });
  it('should properly reject a badly-formatted ADDR message', function() {
    assert.equal(n.parseAddrMessage(new Buffer('0102030405', 'hex')), false);
  });
  it('should properly parse an ALERT message', function() {
    var test = n.parseAlertMessage(new Buffer('73010000003766404f00000000b305434f00000000f2030000f1030000001027000048ee00000064000000004653656520626974636f696e2e6f72672f666562323020696620796f7520686176652074726f75626c6520636f6e6e656374696e67206166746572203230204665627275617279004730450221008389df45f0703f39ec8c1cc42c13810ffcae14995bb648340219e353b63b53eb022009ec65e1c1aaeec1fd334c6b684bde2b3f573060d5b70c3a46723326e4e8a4f1', 'hex'));
    assert.notEqual(test, false);
    assert.equal(test.version, 1);
    assert.equal(test.relay_until.getTime(), 1329620535000);
    assert.equal(test.expiration.getTime(), 1329792435000);
    assert.equal(test.uuid, 1010);
    assert.equal(test.cancel, 1009);
    assert.ok(Array.isArray(test.cancel_set));
    assert.equal(test.cancel_set.length, 0);
    assert.equal(test.min_version, 10000);
    assert.equal(test.max_version, 61000);
    assert.ok(Array.isArray(test.subversion_set));
    assert.equal(test.subversion_set.length, 0);
    assert.equal(test.priority, 100);
    assert.equal(test.comment, "");
    assert.equal(test.status_bar, 'See bitcoin.org/feb20 if you have trouble connecting after 20 February');
    assert.ok(Buffer.isBuffer(test.reserved));
    assert.equal(test.reserved.toString('hex'), '');
    assert.equal(test.signature.toString('hex'), '30450221008389df45f0703f39ec8c1cc42c13810ffcae14995bb648340219e353b63b53eb022009ec65e1c1aaeec1fd334c6b684bde2b3f573060d5b70c3a46723326e4e8a4f1');
  });
  it('should properly reject a badly-formatted ALERT message', function() {
    assert.equal(n.parseAlertMessage(new Buffer('0102030405', 'hex')), false);
  });
  it('should properly parse a BLOCK message', function() {
    var genesis = '01000000'+ // version
      '0000000000000000000000000000000000000000000000000000000000000000'+ // previous block
      '3ba3edfd7a7b12b27ac72c3e67768f617fc81bc3888a51323a9fb8aa4b1e5e4a'+ // merkle root
      '29ab5f49'+ // timestamp
      'ffff001d'+ // difficulty
      '1dac2b7c'+ // nonce
      '01'+ // number of transactions
      '01000000'+ // version
      '01'+ // number of ins
      '0000000000000000000000000000000000000000000000000000000000000000'+ // previous transaction hash
      'ffffffff'+ // previous transaction index
      '4d'+ // script size
      '04ffff001d010445'+
      '5468652054696d65732030332f4a616e2f32303039204368616e63656c6c6f72206f6e206272696e6b206f66207365636f6e64206261696c6f757420666f722062616e6b73'+ // 'The Times 03/Jan/2009 Chancellor on brink of second bailout for banks'
      'ffffffff'+ // sequence
      '01'+ // number of outs
      '00f2052a01000000'+ // value (50 BTC)
      '43'+ // script size
      '4104678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef38c4f35504e51ec112de5c384df7ba0b8d578a4c702b6bf11d5fac'+
      '00000000'; // lock
    var test = n.parseBlockMessage(new Buffer(genesis, 'hex'));
    assert.equal(test.version, 1);
    assert.ok(Buffer.isBuffer(test.prev_block));
    assert.equal(test.prev_block.toString('hex'), '0000000000000000000000000000000000000000000000000000000000000000');
    assert.ok(Buffer.isBuffer(test.merkle_root));
    assert.equal(test.merkle_root.toString('hex'), '3ba3edfd7a7b12b27ac72c3e67768f617fc81bc3888a51323a9fb8aa4b1e5e4a');
    assert.equal(test.timestamp.getTime(), 1231006505000);
    assert.equal(test.difficulty, 486604799);
    assert.equal(test.nonce, 2083236893);
    assert.ok(Array.isArray(test.txns));
    assert.equal(test.txns.length, 1);
    assert.ok(Array.isArray(test.txns[0].txIn));
    assert.equal(test.txns[0].txIn.length, 1);
    assert.ok(Buffer.isBuffer(test.txns[0].txIn[0].outPoint.hash));
    assert.equal(test.txns[0].txIn[0].outPoint.hash.toString('hex'), '0000000000000000000000000000000000000000000000000000000000000000');
    assert.equal(test.txns[0].txIn[0].outPoint.index, 0xFFFFFFFF);
    assert.ok(Buffer.isBuffer(test.txns[0].txIn[0].signature));
    assert.equal(test.txns[0].txIn[0].signature.toString('hex'), '04ffff001d0104455468652054696d65732030332f4a616e2f32303039204368616e63656c6c6f72206f6e206272696e6b206f66207365636f6e64206261696c6f757420666f722062616e6b73');
    assert.equal(test.txns[0].txIn[0].sequence, 0xFFFFFFFF);
    assert.ok(Array.isArray(test.txns[0].txOut));
    assert.equal(test.txns[0].txOut.length, 1);
    assert.ok(Buffer.isBuffer(test.txns[0].txOut[0].value));
    assert.equal(test.txns[0].txOut[0].value.toString('hex'), '00f2052a01000000');
    assert.ok(Buffer.isBuffer(test.txns[0].txOut[0].script));
    assert.equal(test.txns[0].txOut[0].script.toString('hex'), '4104678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef38c4f35504e51ec112de5c384df7ba0b8d578a4c702b6bf11d5fac');
    assert.equal(test.txns[0].lock, 0);
  });
  it('should properly reject a badly-formatted GETBLOCKS message', function() {
    assert.equal(n.parseGetblocksMessage(new Buffer('0102030405', 'hex')), false);
  });
  it('should properly reject a badly-formatted GETHEADERS message', function() {
    assert.equal(n.parseGetheadersMessage(new Buffer('0102030405', 'hex')), false);
  });
  it('should properly reject a badly-formatted HEADERS message', function() {
    assert.equal(n.parseHeadersMessage(new Buffer('0102030405', 'hex')), false);
  });
  it('should properly parse an INV message', function() {
    var test = n.parseInvMessage(new Buffer('0101000000f2bc363cb9ddc6d29beef22c91305da25a88cbb7272dc099c33a75baebfc8c14', 'hex'));
    var items = test.items;
    assert.ok(Array.isArray(items));
    assert.equal(items.length, 1);
    assert.equal(items[0].type, 1);
    assert.equal(items[0].hash.toString('hex'), 'f2bc363cb9ddc6d29beef22c91305da25a88cbb7272dc099c33a75baebfc8c14');
  });
  it('should properly reject a badly-formatted INV message', function() {
    assert.equal(n.parseInvMessage(new Buffer('0102030405', 'hex')), false);
  });
  it('should properly parse a NOTFOUND message', function() {
    var test = n.parseNotfoundMessage(new Buffer('0101000000f2bc363cb9ddc6d29beef22c91305da25a88cbb7272dc099c33a75baebfc8c14', 'hex'));
    var items = test.items;
    assert.ok(Array.isArray(items));
    assert.equal(items.length, 1);
    assert.equal(items[0].type, 1);
    assert.equal(items[0].hash.toString('hex'), 'f2bc363cb9ddc6d29beef22c91305da25a88cbb7272dc099c33a75baebfc8c14');
  });
  it('should properly reject a badly-formatted NOTFOUND message', function() {
    assert.equal(n.parseNotfoundMessage(new Buffer('0102030405', 'hex')), false);
  });
  it('should properly parse a PING message', function() {
    var test = n.parsePingMessage(new Buffer('0102030405060708', 'hex'));
    assert.ok(Buffer.isBuffer(test.nonce));
    assert.equal(test.nonce.toString('hex'), '0102030405060708');
  });
  it('should properly reject a badly-formatted PING message', function() {
    assert.equal(n.parsePingMessage(new Buffer('0102030405', 'hex')), false);
  });
  it('should properly parse a TX message', function() {
    var test = n.parseTxMessage(new Buffer('01000000016DBDDB085B1D8AF75184F0BC01FAD58D1266E9B63B50881990E4B40D6AEE3629000000008B483045022100F3581E1972AE8AC7C7367A7A253BC1135223ADB9A468BB3A59233F45BC578380022059AF01CA17D00E41837A1D58E97AA31BAE584EDEC28D35BD96923690913BAE9A0141049C02BFC97EF236CE6D8FE5D94013C721E915982ACD2B12B65D9B7D59E20A842005F8FC4E02532E873D37B96F09D6D4511ADA8F14042F46614A4C70C0F14BEFF5FFFFFFFF02404B4C00000000001976A9141AA0CD1CBEA6E7458A7ABAD512A9D9EA1AFB225E88AC80FAE9C7000000001976A9140EAB5BEA436A0484CFAB12485EFDA0B78B4ECC5288AC00000000', 'hex'));
    assert.equal(test.version, 1);
    assert.ok(Array.isArray(test.txIn));
    assert.equal(test.txIn.length, 1);
    assert.ok(Array.isArray(test.txOut));
    assert.ok(Buffer.isBuffer(test.txIn[0].outPoint.hash));
    assert.equal(test.txIn[0].outPoint.hash.toString('hex'), '6dbddb085b1d8af75184f0bc01fad58d1266e9b63b50881990e4b40d6aee3629');
    assert.equal(test.txIn[0].outPoint.index, 0);
    assert.ok(Buffer.isBuffer(test.txIn[0].signature));
    assert.equal(test.txIn[0].signature.toString('hex'), '483045022100f3581e1972ae8ac7c7367a7a253bc1135223adb9a468bb3a59233f45bc578380022059af01ca17d00e41837a1d58e97aa31bae584edec28d35bd96923690913bae9a0141049c02bfc97ef236ce6d8fe5d94013c721e915982acd2b12b65d9b7d59e20a842005f8fc4e02532e873d37b96f09d6d4511ada8f14042f46614a4c70c0f14beff5');
    assert.equal(test.txIn[0].sequence, 0xFFFFFFFF);
    assert.equal(test.txOut.length, 2);
    assert.equal(test.txOut[0].value, 5000000);
    assert.ok(Buffer.isBuffer(test.txOut[0].script));
    assert.equal(test.txOut[0].script.toString('hex'), '76a9141aa0cd1cbea6e7458a7abad512a9d9ea1afb225e88ac');
    assert.equal(test.txOut[1].value, 3354000000);
    assert.ok(Buffer.isBuffer(test.txOut[1].script));
    assert.equal(test.txOut[1].script.toString('hex'), '76a9140eab5bea436a0484cfab12485efda0b78b4ecc5288ac');
    assert.equal(test.lock, 0);
    assert.ok(Buffer.isBuffer(test.raw));
    assert.equal(test.hash.toString('hex'), 'e293cdbee28102c38ee58814466c666146b3fad10505cfb4ace77eab513fa7d4');
  })
  it('should properly reject a badly-formatted TX message', function() {
    assert.equal(n.parseTxMessage(new Buffer('0102030405', 'hex')), false);
  });
  it('should properly parse a VERSION message', function() {
    var test = n.parseVersionMessage(new Buffer('7111010001000000000000006017d65200000000010000000000000000000000000000000000ffff4581caaac2e7010000000000000000000000000000000000ffffb78d744f208d351351d2b1e1b0400f2f5361746f7368693a302e382e352fec470400', 'hex'));
    assert.equal(test.version, 70001);
    assert.equal(test.services.readUInt32LE(0), 1);
    assert.equal(test.time.getTime(), 1389762400000);
    assert.equal(test.addr_recv.services.readUInt32LE(0), 1);
    assert.equal(test.addr_recv.host, '69.129.202.170');
    assert.equal(test.addr_recv.port, 49895);
    assert.equal(test.addr_from.services.readUInt32LE(0), 1);
    assert.equal(test.addr_from.host, '183.141.116.79');
    assert.equal(test.addr_from.port, 8333);
    assert.equal(test.nonce.toString('hex'),  '351351d2b1e1b040');
    assert.equal(test.client, '/Satoshi:0.8.5/');
    assert.equal(test.height, 280556);
  });
  it('should properly reject a badly-formatted VERSION message', function() {
    assert.equal(n.parseVersionMessage(new Buffer('0102030405', 'hex')), false);
  });
});


lib/Struct.js
var Struct = exports.Struct = function Struct(raw) {
  Object.defineProperty(this, 'buffer', {
    enumerable: true,
    value: new Buffer(raw.length)
  });
  raw.copy(this.buffer);
  
  this.pointer = 0;
  this.hasFailed = false;
};

Struct.prototype.pointerCheck = function pointerCheck(num) {
  num = +num || 0;
  if (this.buffer.length < this.pointer+num) {
    this.hasFailed = true;
    return false;
  }
};

Struct.prototype.incrPointer = function incrPointer(amount) {
  if (this.hasFailed) return false;
  this.pointer += amount;
  this.pointerCheck();
};

Struct.prototype.setPointer = function setPointer(amount) {
  if (this.hasFailed) return false;
  this.pointer = amount;
  this.pointerCheck();
};

Struct.prototype.readInt8 = function readInt8() {
  if (this.hasFailed || this.pointerCheck() === false) return false;
  var out = this.buffer[this.pointer];
  this.incrPointer(1);
  return out;
};

Struct.prototype.readUInt16LE = function readUInt16LE() {
  if (this.hasFailed || this.pointerCheck(2) === false) return false;
  var out = this.buffer.readUInt16LE(this.pointer);
  this.incrPointer(2);
  return out
};

Struct.prototype.readUInt32LE = function readUInt32LE() {
  if (this.hasFailed || this.pointerCheck(4) === false) return false;
  var out = this.buffer.readUInt32LE(this.pointer);
  this.incrPointer(4);
  return out
};

Struct.prototype.readVarInt = function readVarInt() {
  if (this.hasFailed || this.pointerCheck() === false) return false;
  var flag = this.readInt8();
  if (flag < 0xfd) {
    return flag;
  } else if (flag == 0xfd) {
    return this.readUInt16LE();
  } else if (flag == 0xfe) {
    return this.readUInt32LE();
  } else {
    return this.raw(8);
  }
};

Struct.prototype.readVarString = function readVarString() {
  if (this.hasFailed || this.pointerCheck() === false) return false;
  var length = this.readVarInt();
  var str = [];
  for (var i = 0; i < length; i++) {
    str.push(String.fromCharCode(this.readInt8()));
  }
  return str.join('');
}

Struct.prototype.raw = function raw(length) {
  if (this.hasFailed || this.pointerCheck(length) === false) return false;
  var out = new Buffer(length);
  this.buffer.copy(out, 0, this.pointer, this.pointer+length);
  this.incrPointer(length);
  return out;
}

examples/app.js
var BTCNetwork = require('../lib/BTCNetwork').BTCNetwork;
var dns = require('dns');

var n = new BTCNetwork();

process.once('SIGINT', function() {
  console.log('Got SIGINT; closing...');
  process.once('SIGINT', function() {
    // Double SIGINT; force-kill
    process.exit(0);
  });
  n.shutdown();
});

// Error messages of various severity, from the PeerManager
n.on('error', function error(d) {
  console.log('('+d.severity+'): '+d.message);
});

n.on('status', function status(d) {
  console.log('PeerManager status:', d);
});

// Every message, from every active peer. Gets the message in its raw form (bubble up of 'message' from Peer)
n.on('message', function peerMessage(d) {
  console.log(d.peer.getUUID()+': message', d.command, d.data);
});

// Specific message from the BTCManager. Gets the message in its parsed form
n.on('versionMessage', function versionMessage(d) {
  console.log('VERSION:', d);
});

n.on('transactionInv', function transactionInv(d) {
  console.log('Peer '+d.peer.getUUID()+' knows of Transaction '+d.hash.toString('hex'));
  n.getData({type:1, hash:d.hash}, d.peer, function(err, rs) {
    if (err !== false) {
      console.log('Data returned error: '+err);
      return;
    }
    console.log('Data returned:', rs);
  });
});
n.on('transactionBlock', function transactionInv(d) {
  console.log('Peer '+d.peer.getUUID()+' knows of Block '+d.hash.toString('hex'));
});

/*
// Default launch, with DNS seeds
n.launch();
*/

// Single launch
n.options.minPeers = 1;
n.options.maxPeers = 1;
dns.resolve4('dnsseed.bluematt.me', function(err, addrs) {
  if (err) {
    console.log(err);
    return;
  }
  n.launch(addrs.shift());
  n.manager.addPool(addrs);
});


lib/Message.js
var sha256 = require('crypto-hashing').sha256;

var Message = exports.Message = function Message() {
  this.buffer = new Buffer(10000);
  this.cursor = 0;
};
Message.prototype.checksum = function checksum() {
  return new Buffer(sha256.x2(this.buffer.slice(0, this.cursor)));
};
Message.prototype.raw = function raw() {
  var out = new Buffer(this.cursor);
  this.buffer.copy(out, 0, 0, this.cursor);
  return out;
};
Message.prototype.pad = function pad(num) {
  var data = new Buffer(num);
  data.fill(0);
  return this.put(data);
};
Message.prototype.put = function put(data) {
  if (typeof data == 'number' && data <= 255) {
    this.buffer[this.cursor] = data;
    this.cursor += 1;
    return this;
  }
  
  data.copy(this.buffer, this.cursor);
  this.cursor += data.length;
  return this;
};
Message.prototype.putInt8 = function putInt8(num) {
  if (Buffer.isBuffer(num)) {
    return this.put(num.slice(0, 1));
  }
  return this.put(num);
}
Message.prototype.putInt16 = function putInt16(num) {
  if (Buffer.isBuffer(num)) {
    return this.put(num.slice(0, 2));
  }
  var data = new Buffer(2);
  data.writeUInt16LE(num, 0);
  return this.put(data);
};
Message.prototype.putInt32 = function putInt32(num) {
  if (Buffer.isBuffer(num)) {
    return this.put(num.slice(0, 4));
  } else if (typeof num.getMonth === 'function') {
    return this.putInt32(num.getTime()/1000); // Pull timestamp from Date object
  }
  var data = new Buffer(4);
  data.writeUInt32LE(num, 0);
  return this.put(data);
};
Message.prototype.putInt64 = function putInt64(num) {
  if (Buffer.isBuffer(num)) {
    return this.put(num.slice(0, 8));
  } else if (typeof num.getMonth === 'function') {
    return this.putInt64(num.getTime()/1000); // Pull timestamp from Date object
  }
  // Pad a 32-bit number to fit in a 64-bit space
  var data = new Buffer(8);
  data.fill(0);
  data.writeUInt32LE(num, 0);
  return this.put(data);
};
Message.prototype.putString = function putString(str) {
  var data = new Buffer(str.length);
  for(var i = 0; i < str.length; i++) {
    data[i] = str.charCodeAt(i);
  }
  return this.put(data);
};
Message.prototype.putVarInt = function putVarInt(num) {
  if (num < 0xfd) {
    return this.put(num);
  } else if (num <= 0xffff) {
    return this.put(0xfd).putInt16(num);
  } else if (num <= 0xffffffff) {
    return this.put(0xfe).putInt32(num);
  } else {
    return this.put(0xff).putInt64(num);
  }
};
Message.prototype.varIntSize = function varIntSize(num) {
  if (num < 0xfd) {
    return 1
  } else if (num <= 0xffff) {
    return 3
  } else if (num <= 0xffffffff) {
    return 5
  } else {
    return 9
  }
};
Message.prototype.putVarString = function putVarString(str) {
  return this.putVarInt(str.length).putString(str);
};

Message.prototype.getVarInt = function getVarInt(buff, offset) {
  var flag = buff[offset];
  if (flag < 0xfd) {
    return [flag, 1];
  } else if (flag == 0xfd) {
    return [buff.readUInt16LE(offset+1), 3];
  } else if (flag == 0xfe) {
    return [buff.readUInt32LE(offset+1), 5];
  } else {
    return [buff.slice(offset+1, offset+9), 9];
  }
};
Message.prototype.getVarString = function getVarString(buff, offset) {
  var len = this.getVarInt(buff, offset);
  var str = [];
  for (var i = 0; i < len[0]; i++) {
    str.push(String.fromCharCode(buff[offset+len[1]+i]));
  }
  return str.join('');
};